### PR TITLE
Perform downsampling in AssemblyRegionWalkerSpark's strict mode

### LINKED ISSUE #5476: HaplotypeCallerSpark Enable downsampling in the AssemblyRegionWalker
After #5416 one of the lingering potential sources of difference between HaplotypeCallerSpark and HaplotypeCaller are in the downsampling, which could cost Spark both correctness and performance at pathological sites. This likely requires #5437 or some equivalent change to be implemented so we can save ourselves from materializing and shuffling all the reads in their AssemblyRegions.

### PR DESCRIPTION
**Pull request by tomwhite**
Fixes #5476. This change requires #5437 to work properly.

Not sure how to test this yet.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java [hunk 1]
```diff
@@ -167,8 +167,14 @@
         // 5. Fill in the reads. Each shard is an assembly region, with its overlapping reads.
         JavaRDD<Shard<GATKRead>> assemblyRegionShardedReads = SparkSharder.shard(ctx, reads, GATKRead.class, header.getSequenceDictionary(), assemblyRegionBoundaries, shardingArgs.readShardSize);
 
-        // 6. Convert shards to assembly regions.
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java]: It looks like you haven't reenabled the downsampling in the first step of the pipeline?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java [hunk 2]
```diff
@@ -225,12 +231,16 @@ public ReadlessAssemblyRegion apply(@Nullable AssemblyRegion input) {
                         });
     }
 
-    private static AssemblyRegion toAssemblyRegion(Shard<GATKRead> shard, SAMFileHeader header) {
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java]: Has this affected any of the test outputs? Could we try building a test where we reduce the number of reads per start position to 1 just to demonstrate this is behaving consistently?

Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java]: Has this affected any of the test outputs? Could we try building a test where we reduce the number of reads per start position to 1 just to demonstrate this is behaving consistently?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java [hunk 3]
```diff
@@ -167,8 +167,14 @@
         // 5. Fill in the reads. Each shard is an assembly region, with its overlapping reads.
         JavaRDD<Shard<GATKRead>> assemblyRegionShardedReads = SparkSharder.shard(ctx, reads, GATKRead.class, header.getSequenceDictionary(), assemblyRegionBoundaries, shardingArgs.readShardSize);
 
-        // 6. Convert shards to assembly regions.
-        JavaRDD<AssemblyRegion> assemblyRegions = assemblyRegionShardedReads.map((Function<Shard<GATKRead>, AssemblyRegion>) shard -> toAssemblyRegion(shard, header));
+        // 6. Convert shards to assembly regions. Reads downsampling is done again here, and is assumed to be consistent
+        // with the downsampling done in step 1, since it is deterministic by locus.
+        JavaRDD<AssemblyRegion> assemblyRegions = assemblyRegionShardedReads.mapPartitions((FlatMapFunction<Iterator<Shard<GATKRead>>, AssemblyRegion>) shardedReadIterator -> {
+            final ReadsDownsampler readsDownsampler = assemblyRegionArgs.maxReadsPerAlignmentStart > 0 ?
+                    new PositionalDownsampler(assemblyRegionArgs.maxReadsPerAlignmentStart, header) : null;
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java]: When #5437 goes in, you want to build this positional downsampler with the deterministic arguments enabled.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java (hunk 1 above)
Comment (tomwhite) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java]: It already happens in `getActivityProfileStatesFunction`.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java (hunk 2 above)
Comment (tomwhite) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java]: No, it hasn't affected the tests. Will need to build something like you suggest.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java (hunk 3 above)
Comment (tomwhite) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java]: Right.

**Comment by tomwhite** [comment]
Superceded by #5721

**Pull request closed without merging**